client/buildpacks.go

```diff
@@ -0,0 +1,183 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"time"
+
+	"github.com/buildpacks/pack"
+)
+
+const (
+	image = "conformance-test-func"
+)
+
+type buildpacksFunctionServer struct {
+	output    string
+	source    string
+	target    string
+	funcType  string
+	runtime   string
+	tag       string
+	ctID      string
+	logStdout *os.File
+	logStderr *os.File
+}
+
+func (b *buildpacksFunctionServer) Start() (func(), error) {
+	typ := *functionType
+	if typ == "legacyevent" {
+		typ = "event"
+	}
+
+	ctx := context.Background()
+	if err := b.build(ctx); err != nil {
+		return nil, fmt.Errorf("building function container: %v", err)
+	}
+
+	shutdown, err := b.run()
+	if err != nil {
+		return nil, fmt.Errorf("running function container: %v", err)
+	}
+
+	return shutdown, nil
+}
+
+func (b *buildpacksFunctionServer) OutputFile() ([]byte, error) {
+	cmd := exec.Command("docker", "cp", fmt.Sprintf("%s:/workspace/%s", b.containerID(), b.output), ".")
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return nil, fmt.Errorf("failed to copy output file from the container: %v: %s", err, string(output))
+	}
+	return ioutil.ReadFile(b.output)
+}
+
+func (b *buildpacksFunctionServer) build(ctx context.Context) error {
+	builder := fmt.Sprintf("us.gcr.io/fn-img/buildpacks/%s/builder:%s", b.runtime, b.tag)
+	packClient, err := pack.NewClient()
+	if err != nil {
+		return fmt.Errorf("getting pack client: %v", err)
+	}
+	err = packClient.Build(ctx, pack.BuildOptions{
+		Image:    image,
+		Builder:  builder,
+		AppPath:  b.source,
+		Registry: "",
+		Env: map[string]string{
+			"GOOGLE_FUNCTION_SOURCE":         b.source,
```

Comment:
Not all buildpacks support GOOGLE_FUNCTION_SOURCE. The Node.js one blocks this env var. Also, this can be different from `AppPath`, which is the path to the root directory of the function source code (that `pack` uses to build the docker image). Maybe remove this for now?